src/Plugin/InvokeStrategy/OnEventStrategy.php

```diff
@@ -27,6 +27,10 @@ function (ActionEvent $actionEvent): void {
                 $message = $actionEvent->getParam(MessageBus::EVENT_PARAM_MESSAGE);
                 $handlers = $actionEvent->getParam(EventBus::EVENT_PARAM_EVENT_LISTENERS, []);
 
+                if ($actionEvent->getParam(MessageBus::EVENT_PARAM_MESSAGE_HANDLED, false)) {
```

Comment:
move this code up, message and handlers don't need to get fetched before this check.